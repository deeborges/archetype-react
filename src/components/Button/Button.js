import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ButtonColors = {
	default: 'default',
	primary: 'primary',
	danger: 'danger',
};

const ButtonVariants = {
	default: 'default',
	outlined: 'outlined',
	link: 'link',
};

const getMainColor = ({ theme, color }) => {
	switch (color) {
		case ButtonColors.primary:
			return theme.colors.blue;
		case ButtonColors.danger:
			return theme.colors.red;
		default:
			return '#e0e0e0';
	}
};

const getColorText = (props) => props.theme.colors.purple;
const getOutlinedColorText = (props) => {
	if (props.color == ButtonColors.default) {
		return '#212121';
	}
	return getMainColor(props);
};
const getLinkColorText = (props) => {
	if (props.color == ButtonColors.default) {
		return '#757575';
	}
	return getMainColor(props);
};

export const Button = styled.button`
	font-size: 1rem;
	font-weight: 600;
	text-transform: uppercase;
	padding: 12px 36px;
	cursor: pointer;
	background-color: ${getMainColor};
	border: 2px solid ${getMainColor};
	color: ${getColorText};

	&:disabled {
		opacity: 0.6;
		cursor: not-allowed;
	}

	&:hover:enabled {
		background-color: '#c3c3c3';
	}
`;

const ButtonOutlined = styled(Button)`
	background-color: transparent;
	color: ${getOutlinedColorText};

	&:hover:enabled {
		background-color: transparent;
		color: ${getOutlinedColorText};
	}
`;

const ButtonLink = styled(Button)`
	background-color: transparent;
	border-color: transparent;
	color: ${getLinkColorText};
	padding-left: 0;
	padding-right: 0;

	&:hover:enabled {
		background-color: transparent;
		opacity: 0.7;
	}
`;

const ButtonWrapper = (props) => {
	switch (props.variant) {
		case ButtonVariants.outlined:
			return <ButtonOutlined {...props} />;
		case ButtonVariants.link:
			return <ButtonLink {...props} />;
		default:
			return <Button {...props} />;
	}
};

ButtonWrapper.defaultProps = {
	type: 'button',
	children: undefined,
	color: 'default',
	variant: 'default',
};

ButtonWrapper.propTypes = {
	type: PropTypes.string,
	children: PropTypes.node,
	color: PropTypes.oneOf(Object.values(ButtonColors)),
	variant: PropTypes.oneOf(Object.values(ButtonVariants)),
};

export default ButtonWrapper;
