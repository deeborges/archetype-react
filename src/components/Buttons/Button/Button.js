import styled from 'styled-components';
import PropTypes from 'prop-types';

const ButtonColors = {
  danger: 'danger',
  default: 'deafult',
  primary: 'primary'
};

export const Button = styled.button`
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ffc107;
  border: 2px solid #ffc107;
  color: ${(props) => props.theme.colors.yellow};
`;

Button.defaultProps = {
  children: undefined,
  color: 'default',
  type: 'button'
};

Button.propTypes = {
  children: PropTypes.node,
  color: PropTypes.oneOf(Object.values(ButtonColors)),
  type: PropTypes.string
};
