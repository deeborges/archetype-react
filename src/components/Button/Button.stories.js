import React from 'react';
import Button from './index';

export default {
  title: 'Components/Button',
  component: Button
};

export const usage = () => <Button>Meu botão</Button>;
