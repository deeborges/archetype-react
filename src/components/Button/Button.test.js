import React from 'react';
import { render, fireEvent } from '../../test-utils.js';
import Button, { ButtonColors, ButtonVariants } from './Button';

test('renderizar um botao com um texto', () => {
	const { getByText } = render(<Button>botão de teste</Button>);
	expect(getByText('botão de teste')).toBeInTheDocument();
});

test('disparar um evento ao clickar', () => {
	const handleClick = jest.fn();

	const { getByRole } = render(
		<Button onClick={handleClick}>Click here</Button>
	);
	fireEvent.click(getByRole('button'));
	expect(handleClick).toBeCalled();
});

test.each(Object.values(ButtonColors).map((item) => [item]))(
	'renderizar com cor %s',
	(color) => {
		const { asFragment } = render(
			<Button color={color}>botão de teste</Button>
		);
		expect(asFragment()).toMatchSnapshot();
	}
);

test.each(Object.values(ButtonVariants).map((item) => [item]))(
	'renderizar a variacao %s',
	(variant) => {
		const { asFragment } = render(
			<Button variant={variant}>botão de teste</Button>
		);
		expect(asFragment()).toMatchSnapshot();
	}
);

test('renderizar com cor primary e variant outlined', () => {
	const { asFragment } = render(
		<Button color="primary" variant="outlined">
			botão de teste
		</Button>
	);
	expect(asFragment()).toMatchSnapshot();
});
