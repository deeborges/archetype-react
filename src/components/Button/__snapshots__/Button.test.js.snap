// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderizar a variacao default 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ jIFCMh"
    color="default"
    type="button"
  >
    botão de teste
  </button>
</DocumentFragment>
`;

exports[`renderizar a variacao link 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ sc-dlfnbm jIFCMh dGBDFc"
    color="default"
    type="button"
  >
    botão de teste
  </button>
</DocumentFragment>
`;

exports[`renderizar a variacao outlined 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ sc-gsTCUz jIFCMh gqYILX"
    color="default"
    type="button"
  >
    botão de teste
  </button>
</DocumentFragment>
`;

exports[`renderizar com cor danger 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ hIutWm"
    color="danger"
    type="button"
  >
    botão de teste
  </button>
</DocumentFragment>
`;

exports[`renderizar com cor default 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ jIFCMh"
    color="default"
    type="button"
  >
    botão de teste
  </button>
</DocumentFragment>
`;

exports[`renderizar com cor primary 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ homhHy"
    color="primary"
    type="button"
  >
    botão de teste
  </button>
</DocumentFragment>
`;

exports[`renderizar com cor primary e variant outlined 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ sc-gsTCUz homhHy gOirBP"
    color="primary"
    type="button"
  >
    botão de teste
  </button>
</DocumentFragment>
`;
